File: lib/anilibria.py
import os
import re

import requests


API_BASE_URL = "https://aniliberty.top/api/v1"
TORRENTS_PAGE_URLS = [
    "https://www.anilibria.top/anime/torrents",
    "https://aniliberty.top/anime/torrents",
    "https://anilibria.top/anime/torrents",
]
DEFAULT_HEADERS = {
    "User-Agent": "workspace-gojo-satoru/1.0 (+https://aniliberty.top)",
    "Accept": "application/json, text/html;q=0.9, */*;q=0.8",
}


def get_anilibria_proxy_url():
    return (os.getenv("ANILIBERTY_PROXY_URL") or "").strip()


def _build_request_kwargs(timeout, params=None):
    kwargs = {
        "timeout": timeout,
        "headers": DEFAULT_HEADERS,
    }
    if params is not None:
        kwargs["params"] = params

    proxy_url = get_anilibria_proxy_url()
    if proxy_url:
        kwargs["proxies"] = {
            "http": proxy_url,
            "https": proxy_url,
        }
    return kwargs


def _extract_names(payload):
    values = []

    names = payload.get("names") or payload.get("name") or {}
    if isinstance(names, dict):
        for key in ["ru", "en", "english", "alternative", "main"]:
            value = names.get(key)
            if isinstance(value, str) and value.strip():
                values.append(value.strip())

    title = payload.get("title")
    if isinstance(title, str) and title.strip():
        values.append(title.strip())

    alias = payload.get("alias")
    if isinstance(alias, str) and alias.strip():
        values.append(alias.strip())

    deduped = []
    seen = set()
    for value in values:
        lowered = value.lower()
        if lowered in seen:
            continue
        seen.add(lowered)
        deduped.append(value)
    return deduped


def _normalize_skip_interval(skip_type, payload):
    if not isinstance(payload, dict):
        return None

    start = (
        payload.get("start")
        or payload.get("from")
        or payload.get("startTime")
        or payload.get("start_time")
    )
    end = (
        payload.get("end")
        or payload.get("to")
        or payload.get("stop")
        or payload.get("endTime")
        or payload.get("end_time")
    )
    if start is None or end is None:
        return None

    try:
        return {
            "type": skip_type,
            "start": float(start),
            "end": float(end),
            "source": "anilibria_exact",
            "confidence": "high",
        }
    except (TypeError, ValueError):
        return None


def _collect_skip_segments(payload):
    if not isinstance(payload, dict):
        return []

    candidates = []
    skips = payload.get("skips")
    if isinstance(skips, dict):
        candidates.append(skips)
    candidates.append(payload)

    segments = []
    for item in candidates:
        for skip_type, aliases in {
            "op": ["op", "opening"],
            "ed": ["ed", "ending"],
        }.items():
            for alias in aliases:
                if alias not in item:
                    continue
                segment = _normalize_skip_interval(skip_type, item.get(alias))
                if segment:
                    segments.append(segment)
                    break

    deduped = {}
    for segment in segments:
        deduped[(segment["type"], segment["start"], segment["end"])] = segment
    return sorted(deduped.values(), key=lambda item: (item["start"], item["type"]))


def _match_title(payload, title, aliases):
    wanted = {title.strip().lower()}
    wanted.update(
        alias.strip().lower()
        for alias in aliases
        if isinstance(alias, str) and alias.strip()
    )
    available = {name.lower() for name in _extract_names(payload)}
    return bool(wanted & available)


def _request(url, params=None):
    response = requests.get(url, **_build_request_kwargs(timeout=20, params=params))
    response.raise_for_status()
    return response.json(), response.url


def _request_text(url, params=None):
    response = requests.get(url, **_build_request_kwargs(timeout=30, params=params))
    response.raise_for_status()
    return response.text, response.url


def _normalize_release_list(data):
    releases = []
    for candidate in _iter_release_candidates(data):
        if isinstance(candidate, dict):
            releases.append(candidate)
    return releases


def _extract_release_aliases_from_torrents_page(html):
    aliases = []
    seen = set()
    patterns = [
        r"/anime/releases/release/([^/\"'?]+)",
        r"/anime/releases/(?!release/)([^/\"'?]+)",
    ]
    for pattern in patterns:
        for alias in re.findall(pattern, html or ""):
            normalized = alias.strip()
            if not normalized or normalized in seen:
                continue
            seen.add(normalized)
            aliases.append(normalized)
    return aliases


def _build_recent_releases_from_torrents_page(limit, urls, errors):
    aliases = []
    for page_url in TORRENTS_PAGE_URLS:
        try:
            html, request_url = _request_text(page_url)
            urls.append(str(request_url))
            aliases = _extract_release_aliases_from_torrents_page(html)
            if aliases:
                break
        except requests.RequestException as exc:
            errors.append(f"torrents_page: {exc}")

    if not aliases:
        return []

    releases = []
    for alias in aliases:
        if len(releases) >= int(limit):
            break
        try:
            release_details = get_release_details(alias)
            urls.append(release_details["request_url"])
            release_payload = release_details["release"]
            if isinstance(release_payload, dict):
                releases.append(release_payload)
        except requests.RequestException as exc:
            errors.append(f"anime/releases/{alias}: {exc}")

    return releases


def _build_recent_releases_from_api(limit, urls, errors):
    api_attempts = [
        ("anime/releases/latest", {"limit": int(limit)}),
        ("anime/releases/latest", None),
    ]

    for path, params in api_attempts:
        try:
            data, request_url = _request(f"{API_BASE_URL}/{path}", params=params)
            urls.append(str(request_url))
        except requests.RequestException as exc:
            errors.append(f"{path}: {exc}")
            continue

        releases = _normalize_release_list(data)
        if releases:
            return releases

    return []


def list_recent_releases(limit=25):
    urls = []
    errors = []
    releases = _build_recent_releases_from_api(limit, urls, errors)
    if not releases:
        releases = _build_recent_releases_from_torrents_page(limit, urls, errors)

    if not releases:
        error = "; ".join(errors) if errors else "no_releases_found"
        raise RuntimeError(f"AniLibria recent releases lookup failed: {error}")

    def sort_key(item):
        return (
            item.get("fresh_at")
            or item.get("updated_at")
            or item.get("created_at")
            or ""
        )

    releases.sort(key=sort_key, reverse=True)
    return {
        "releases": releases[: int(limit)],
        "request_urls": urls,
    }


def _iter_release_candidates(data):
    if isinstance(data, list):
        return data
    if isinstance(data, dict):
        for key in ["data", "items", "list", "releases"]:
            value = data.get(key)
            if isinstance(value, list):
                return value
        return [data]
    return []


def _find_release(title, season, aliases):
    urls = []
    errors = []
    candidates = []

    search_attempts = [
        ("app/search/releases", {"query": title}),
        ("anime/releases/list", {"aliases": ",".join(filter(None, [title, *aliases]))}),
    ]

    for path, params in search_attempts:
        try:
            data, request_url = _request(f"{API_BASE_URL}/{path}", params=params)
            urls.append(str(request_url))
        except requests.RequestException as exc:
            errors.append(f"{path}: {exc}")
            continue

        candidates.extend(_iter_release_candidates(data))

    for candidate in candidates:
        if not isinstance(candidate, dict):
            continue
        if not _match_title(candidate, title, aliases):
            continue
        if season is not None:
            season_value = candidate.get("season_number")
            if season_value is not None and str(season_value) != str(season):
                continue
        release_id = candidate.get("id") or candidate.get("release_id")
        release_alias = candidate.get("alias")
        if release_id is not None or release_alias:
            return candidate, urls, None

    error = "; ".join(errors) if errors else "release_not_found"
    return None, urls, error


def _get_release_payload(id_or_alias):
    data, request_url = _request(f"{API_BASE_URL}/anime/releases/{id_or_alias}")
    return data, str(request_url)


def get_release_details(id_or_alias):
    payload, request_url = _get_release_payload(id_or_alias)
    return {
        "release": payload,
        "request_url": request_url,
    }


def _find_episode_payload(release_payload, episode_number):
    episodes = release_payload.get("episodes")
    if not isinstance(episodes, list):
        return None

    for candidate in episodes:
        if not isinstance(candidate, dict):
            continue
        candidate_number = (
            candidate.get("number")
            or candidate.get("episode")
            or candidate.get("ordinal")
        )
        if candidate_number is None:
            continue
        try:
            if int(candidate_number) == int(episode_number):
                return candidate
        except (TypeError, ValueError):
            continue
    return None


def get_anilibria_segments(title, season, episode_number, source=None, aliases=None):
    aliases = aliases or []
    request_urls = []

    release_stub, release_urls, release_error = _find_release(title, season, aliases)
    request_urls.extend(release_urls)
    if release_stub is None:
        return {
            "segments": [],
            "request_error": f"AniLibria release lookup failed: {release_error}",
            "request_urls": request_urls,
            "provider": "anilibria",
        }

    release_id_or_alias = release_stub.get("alias") or release_stub.get("id") or release_stub.get("release_id")
    try:
        release_details = get_release_details(release_id_or_alias)
        release_payload = release_details["release"]
        request_urls.append(release_details["request_url"])
    except requests.RequestException as exc:
        return {
            "segments": [],
            "request_error": f"AniLibria release details failed: {exc}",
            "request_urls": request_urls,
            "provider": "anilibria",
        }

    episode_payload = _find_episode_payload(release_payload, episode_number)
    if episode_payload is None:
        return {
            "segments": [],
            "request_error": "AniLibria episode lookup failed: episode_not_found",
            "request_urls": request_urls,
            "provider": "anilibria",
        }

    segments = _collect_skip_segments(episode_payload)
    if not segments:
        segments = _collect_skip_segments(release_payload)

    return {
        "segments": segments,
        "request_error": None if segments else "AniLibria returned no skip data",
        "request_urls": request_urls,
        "provider": "anilibria",
    }

File: src/api/anilibria.ts
import { request as httpRequest } from "node:http";
import { request as httpsRequest } from "node:https";
import { SocksProxyAgent } from "socks-proxy-agent";
import type { ReleasePayload, ReleaseEpisode, ReleaseVariant, AniLibriaResult, Segment } from "../shared/types";

const API_BASE_URL = "https://aniliberty.top/api/v1";
const TORRENTS_PAGE_URLS = [
  "https://www.anilibria.top/anime/torrents",
  "https://aniliberty.top/anime/torrents",
  "https://anilibria.top/anime/torrents",
];
const DEFAULT_HEADERS = {
  "User-Agent": "workspace-gojo-satoru/2.0 (+https://aniliberty.top)",
  "Accept": "application/json, text/html;q=0.9, */*;q=0.8",
};

function getProxyUrl(): string | null {
  return Bun.env.ANILIBERTY_PROXY_URL?.trim() || null;
}

function createAgent(url: string): unknown | undefined {
  const proxy = getProxyUrl();
  if (!proxy) return undefined;

  const parsed = new URL(url);
  // Only use proxy agent for HTTPS targets with SOCKS proxy
  if (proxy.startsWith("socks") && parsed.protocol === "https:") {
    return new SocksProxyAgent(proxy);
  }
  return undefined;
}

async function nodeFetch(url: string, timeout = 20): Promise<{ text: string; status: number }> {
  const parsed = new URL(url);
  const agent = createAgent(url) as SocksProxyAgent | undefined;

  return new Promise((resolve, reject) => {
    const req = httpsRequest(
      url,
      {
        headers: DEFAULT_HEADERS,
        agent,
        timeout: timeout * 1000,
      },
      (res) => {
        let data = "";
        res.on("data", chunk => data += chunk);
        res.on("end", () => resolve({ text: data, status: res.statusCode ?? 0 }));
        res.on("error", reject);
      },
    );
    req.on("error", reject);
    req.on("timeout", () => { req.destroy(); reject(new Error(`Request timeout: ${url}`)); });
    req.end();
  });
}

async function requestJson(url: string, timeout = 20): Promise<{ data: unknown; requestUrl: string }> {
  const { text, status } = await nodeFetch(url, timeout);
  if (status < 200 || status >= 300) throw new Error(`AniLibria HTTP ${status} for ${url}`);
  return { data: JSON.parse(text), requestUrl: url };
}

async function requestText(url: string, timeout = 30): Promise<{ text: string; requestUrl: string }> {
  const { text, status } = await nodeFetch(url, timeout);
  if (status < 200 || status >= 300) throw new Error(`AniLibria HTTP ${status} for ${url}`);
  return { text, requestUrl: url };
}

// ============================================================================
// Recent releases
// ============================================================================

function normalizeReleaseList(data: unknown): ReleasePayload[] {
  if (Array.isArray(data)) return data as ReleasePayload[];
  if (typeof data === "object" && data !== null) {
    const obj = data as Record<string, unknown>;
    for (const key of ["data", "items", "list", "releases"]) {
      if (Array.isArray(obj[key])) return obj[key] as ReleasePayload[];
    }
    return [data as ReleasePayload];
  }
  return [];
}

async function buildRecentReleasesFromApi(limit: number, urls: string[], errors: string[]): Promise<ReleasePayload[]> {
  const attempts: [string, string | null][] = [
    [`${API_BASE_URL}/anime/releases/latest`, `limit=${limit}`],
    [`${API_BASE_URL}/anime/releases/latest`, null],
  ];

  for (const [path, query] of attempts) {
    try {
      const url = query ? `${path}?${query}` : path;
      const { data, requestUrl } = await requestJson(url);
      urls.push(requestUrl);
      const releases = normalizeReleaseList(data);
      if (releases.length > 0) return releases;
    } catch (err) {
      errors.push(`${path}: ${err}`);
    }
  }
  return [];
}

function extractReleaseAliasesFromPage(html: string): string[] {
  const aliases: string[] = [];
  const seen = new Set<string>();
  const patterns = [/\/anime\/releases\/release\/([^/"'?]+)/g, /\/anime\/releases\/(?!release\/)([^/"'?]+)/g];
  for (const pattern of patterns) {
    for (const match of html.matchAll(pattern)) {
      const alias = match[1]!.trim();
      if (!seen.has(alias)) { seen.add(alias); aliases.push(alias); }
    }
  }
  return aliases;
}

async function buildRecentReleasesFromPage(limit: number, urls: string[], errors: string[]): Promise<ReleasePayload[]> {
  let aliases: string[] = [];
  for (const pageUrl of TORRENTS_PAGE_URLS) {
    try {
      const { text, requestUrl } = await requestText(pageUrl);
      urls.push(requestUrl);
      aliases = extractReleaseAliasesFromPage(text);
      if (aliases.length > 0) break;
    } catch (err) {
      errors.push(`torrents_page: ${err}`);
    }
  }

  if (aliases.length === 0) return [];

  const releases: ReleasePayload[] = [];
  for (const alias of aliases) {
    if (releases.length >= limit) break;
    try {
      const details = await getReleaseDetails(alias);
      urls.push(details.request_url);
      const releasePayload = details.release;
      if (typeof releasePayload === "object" && releasePayload !== null) {
        releases.push(releasePayload);
      }
    } catch (err) {
      errors.push(`anime/releases/${alias}: ${err}`);
    }
  }
  return releases;
}

export async function listRecentReleases(limit = 50) {
  const urls: string[] = [];
  const errors: string[] = [];

  let releases = await buildRecentReleasesFromApi(limit, urls, errors);
  if (releases.length === 0) {
    releases = await buildRecentReleasesFromPage(limit, urls, errors);
  }

  if (releases.length === 0) {
    const error = errors.length > 0 ? errors.join("; ") : "no_releases_found";
    throw new Error(`AniLibria recent releases lookup failed: ${error}`);
  }

  releases.sort((a, b) => {
    const aDate = a.fresh_at ?? a.updated_at ?? a.created_at ?? "";
    const bDate = b.fresh_at ?? b.updated_at ?? b.created_at ?? "";
    return bDate.localeCompare(aDate);
  });

  return { releases: releases.slice(0, limit), request_urls: urls };
}

// ============================================================================
// Release details
// ============================================================================

async function getReleasePayload(idOrAlias: string | number): Promise<{ data: unknown; requestUrl: string }> {
  return requestJson(`${API_BASE_URL}/anime/releases/${idOrAlias}`);
}

export async function getReleaseDetails(idOrAlias: string | number) {
  const { data, requestUrl } = await getReleasePayload(idOrAlias);
  return { release: data as ReleasePayload, request_url: requestUrl };
}

// ============================================================================
// Release search
// ============================================================================

function extractNames(payload: Record<string, unknown>): string[] {
  const values: string[] = [];
  const names = (payload.names ?? payload.name ?? {}) as Record<string, string>;
  if (typeof names === "object") {
    for (const key of ["ru", "en", "english", "alternative", "main"]) {
      const v = names[key];
      if (typeof v === "string" && v.trim()) values.push(v.trim());
    }
  }
  const title = payload.title;
  if (typeof title === "string" && title.trim()) values.push(title.trim());
  const alias = payload.alias;
  if (typeof alias === "string" && alias.trim()) values.push(alias.trim());

  const seen = new Set<string>();
  return values.filter(v => { const l = v.toLowerCase(); if (seen.has(l)) return false; seen.add(l); return true; });
}

function matchTitle(payload: Record<string, unknown>, title: string, aliases: string[]): boolean {
  const wanted = new Set([title.trim().toLowerCase(), ...aliases.filter(a => typeof a === "string" && a.trim()).map(a => a.trim().toLowerCase())]);
  const available = new Set(extractNames(payload).map(n => n.toLowerCase()));
  for (const w of wanted) { if (available.has(w)) return true; }
  return false;
}

async function findRelease(title: string, season: number | null, aliases: string[]) {
  const urls: string[] = [];
  const errors: string[] = [];
  const candidates: ReleasePayload[] = [];

  const searchAttempts: [string, string][] = [
    [`${API_BASE_URL}/app/search/releases`, `query=${encodeURIComponent(title)}`],
    [`${API_BASE_URL}/anime/releases/list`, `aliases=${encodeURIComponent([title, ...aliases].filter(Boolean).join(","))}`],
  ];

  for (const [path, query] of searchAttempts) {
    try {
      const { data, requestUrl } = await requestJson(`${path}?${query}`);
      urls.push(requestUrl);
      candidates.push(...normalizeReleaseList(data));
    } catch (err) {
      errors.push(`${path}: ${err}`);
    }
  }

  for (const candidate of candidates) {
    if (!matchTitle(candidate as Record<string, unknown>, title, aliases)) continue;
    if (season !== null) {
      const sv = (candidate as Record<string, unknown>).season_number;
      if (sv !== undefined && String(sv) !== String(season)) continue;
    }
    const releaseId = (candidate as Record<string, unknown>).id ?? (candidate as Record<string, unknown>).release_id;
    const releaseAlias = (candidate as Record<string, unknown>).alias as string | undefined;
    if (releaseId !== undefined || releaseAlias) return { release: candidate, urls, error: null };
  }

  return { release: null, urls, error: errors.length > 0 ? errors.join("; ") : "release_not_found" };
}

// ============================================================================
// Episode skips
// ============================================================================

function findEpisodePayload(releasePayload: ReleasePayload, episodeNumber: number): ReleaseEpisode | null {
  const episodes = releasePayload.episodes;
  if (!Array.isArray(episodes)) return null;

  for (const candidate of episodes) {
    const candidateNumber = candidate.number ?? candidate.episode ?? candidate.ordinal;
    if (candidateNumber === undefined) continue;
    if (Number(candidateNumber) === Number(episodeNumber)) return candidate;
  }
  return null;
}

function normalizeSkipInterval(skipType: string, payload: Record<string, unknown>): Segment | null {
  const start = payload.start ?? payload.from ?? payload.startTime ?? payload.start_time;
  const end = payload.end ?? payload.to ?? payload.stop ?? payload.endTime ?? payload.end_time;
  if (start === undefined || end === undefined) return null;
  try {
    return { type: skipType, start: Number(start), end: Number(end), source: "anilibria_exact", confidence: "high" };
  } catch {
    return null;
  }
}

function collectSkipSegments(payload: Record<string, unknown>): Segment[] {
  const candidates: Record<string, unknown>[] = [];
  if (payload.skips && typeof payload.skips === "object") candidates.push(payload.skips as Record<string, unknown>);
  candidates.push(payload);

  const segments: Segment[] = [];
  for (const item of candidates) {
    for (const [skipType, aliases] of [["op", ["op", "opening"]], ["ed", ["ed", "ending"]]] as [string, string[]][]) {
      for (const alias of aliases) {
        if (!(alias in item)) continue;
        const seg = normalizeSkipInterval(skipType, item[alias] as Record<string, unknown>);
        if (seg) { segments.push(seg); break; }
      }
    }
  }

  const deduped = new Map<string, Segment>();
  for (const seg of segments) {
    deduped.set(`${seg.type}|${seg.start}|${seg.end}`, seg);
  }
  return [...deduped.values()].sort((a, b) => a.start - b.start);
}

export async function getAnilibriaSegments(
  title: string,
  season: number,
  episodeNumber: number,
  _source?: Record<string, unknown>,
  aliases: string[] = [],
): Promise<AniLibriaResult> {
  const requestUrls: string[] = [];

  const { release: releaseStub, urls: releaseUrls, error: releaseError } = await findRelease(title, season, aliases);
  requestUrls.push(...releaseUrls);

  if (!releaseStub) {
    return { segments: [], request_error: `AniLibria release lookup failed: ${releaseError}`, request_urls: requestUrls, provider: "anilibria" };
  }

  const releaseIdOrAlias = (releaseStub as Record<string, unknown>).alias ?? (releaseStub as Record<string, unknown>).id ?? (releaseStub as Record<string, unknown>).release_id;
  try {
    const details = await getReleaseDetails(releaseIdOrAlias as string | number);
    const releasePayload = details.release;
    requestUrls.push(details.request_url);

    const episodePayload = findEpisodePayload(releasePayload, episodeNumber);
    if (!episodePayload) {
      return { segments: [], request_error: "AniLibria episode lookup failed: episode_not_found", request_urls: requestUrls, provider: "anilibria" };
    }

    let segments = collectSkipSegments(episodePayload as Record<string, unknown>);
    if (segments.length === 0) {
      segments = collectSkipSegments(releasePayload as Record<string, unknown>);
    }

    return {
      segments,
      request_error: segments.length > 0 ? null : "AniLibria returned no skip data",
      request_urls: requestUrls,
      provider: "anilibria",
    };
  } catch (err) {
    return { segments: [], request_error: `AniLibria release details failed: ${err}`, request_urls: requestUrls, provider: "anilibria" };
  }
}

// ============================================================================
// Release variant extraction (for discovery)
// ============================================================================

export function collectReleaseEpisodeNumbers(releasePayload: ReleasePayload): number[] {
  const episodes = releasePayload.episodes;
  if (!Array.isArray(episodes)) return [];

  const numbers = new Set<number>();
  for (const item of episodes) {
    const num = item.number ?? item.episode ?? item.ordinal;
    if (num === undefined) continue;
    const parsed = Number(num);
    if (Number.isNaN(parsed) || parsed <= 0) continue;
    numbers.add(parsed);
  }
  return [...numbers].sort((a, b) => a - b);
}

function parseVariantCodec(value: unknown): "avc" | "hevc" | null {
  const text = String(value ?? "").trim().toLowerCase();
  if (!text) return null;
  if (["avc", "x264", "h.264", "h264"].some(m => text.includes(m))) return "avc";
  if (["hevc", "x265", "h.265", "h265"].some(m => text.includes(m))) return "hevc";
  return null;
}

function extractVariantCodec(payload: Record<string, unknown>): "avc" | "hevc" | null {
  for (const value of [payload.codec, payload.video_codec, payload.videoCodec, payload.label, payload.title, payload.name]) {
    const codec = parseVariantCodec(value);
    if (codec) return codec;
  }
  return null;
}

function findMagnetValue(payload: unknown): string | null {
  if (typeof payload === "string") {
    const v = payload.trim();
    return v.startsWith("magnet:?") ? v : null;
  }
  if (Array.isArray(payload)) {
    for (const item of payload) { const m = findMagnetValue(item); if (m) return m; }
  }
  if (typeof payload === "object" && payload !== null) {
    for (const value of Object.values(payload)) { const m = findMagnetValue(value); if (m) return m; }
  }
  return null;
}

function parseVariantResolution(payload: Record<string, unknown>): string | null {
  for (const key of ["resolution", "quality", "video_quality", "videoQuality"]) {
    const v = payload[key];
    if (typeof v === "string" && v.trim()) return v.trim();
  }
  return null;
}

function extractVariantLabel(payload: Record<string, unknown>): string | null {
  for (const key of ["label", "quality_label", "qualityLabel", "title", "name"]) {
    const v = payload[key];
    if (typeof v === "string" && v.trim()) return v.trim();
  }
  const codec = extractVariantCodec(payload);
  const resolution = parseVariantResolution(payload);
  const parts = [codec?.toUpperCase(), resolution].filter(Boolean);
  return parts.length > 0 ? parts.join(" ") : null;
}

function extractVariantLabelEpisodeNumbers(label: string | null): number[] | null {
  if (!label) return null;
  const matches = [...label.matchAll(/\[(\d{1,3})(?:\s*-\s*(\d{1,3}))?\]/g)];
  if (matches.length === 0) return null;
  const lastMatch = matches[matches.length - 1]!;
  const start = Number(lastMatch[1]);
  const end = lastMatch[2] ? Number(lastMatch[2]) : start;
  if (start <= 0 || end <= 0 || start > end) return null;
  return Array.from({ length: end - start + 1 }, (_, i) => start + i);
}

function* iterReleaseVariantPayloads(releasePayload: ReleasePayload): Generator<Record<string, unknown>> {
  const seen = new Set<unknown>();
  const variants: unknown[] = [];

  function push(candidate: unknown) {
    if (typeof candidate !== "object" || candidate === null || seen.has(candidate)) return;
    seen.add(candidate);
    variants.push(candidate);
  }

  const torrents = releasePayload.torrents;
  if (Array.isArray(torrents)) { for (const item of torrents) push(item); }
  else if (typeof torrents === "object" && torrents !== null) {
    for (const value of Object.values(torrents)) {
      if (Array.isArray(value)) { for (const item of value) push(item); }
      else push(value);
    }
  }

  for (const key of ["qualities", "quality", "variants", "versions"]) {
    const value = (releasePayload as Record<string, unknown>)[key];
    if (Array.isArray(value)) { for (const item of value) push(item); }
    else if (typeof value === "object" && value !== null) {
      for (const nested of Object.values(value as Record<string, unknown>)) {
        if (Array.isArray(nested)) { for (const item of nested) push(item); }
        else push(nested);
      }
    }
  }

  const torrent = (releasePayload as Record<string, unknown>).torrent;
  if (typeof torrent === "object" && torrent !== null) {
    const legacyMagnet = findMagnetValue(torrent);
    if (legacyMagnet) {
      yield* [{ codec: "avc", label: "legacy", magnet: legacyMagnet, episodes: releasePayload.episodes }];
    }
  }

  yield* variants as Record<string, unknown>[];
}

export function extractReleaseSourceVariants(releasePayload: ReleasePayload): ReleaseVariant[] {
  const variants: ReleaseVariant[] = [];
  const deduped = new Set<string>();
  const releaseEpisodes = collectReleaseEpisodeNumbers(releasePayload);

  for (const candidate of iterReleaseVariantPayloads(releasePayload)) {
    const magnet = findMagnetValue(candidate);
    const codec = extractVariantCodec(candidate);
    const label = extractVariantLabel(candidate);
    const labelEpisodes = extractVariantLabelEpisodeNumbers(label);

    let episodes = releaseEpisodes;
    if (labelEpisodes) {
      const labelSet = new Set(labelEpisodes);
      episodes = releaseEpisodes.filter(e => labelSet.has(e));
    }

    if (!magnet || !codec || episodes.length === 0) continue;

    const identity = `${magnet}|${codec}|${episodes.join(",")}`;
    if (deduped.has(identity)) continue;
    deduped.add(identity);

    variants.push({
      codec,
      resolution: parseVariantResolution(candidate),
      magnet,
      available_episodes: episodes,
      label: label ?? undefined,
    });
  }

  return variants;
}

export function selectReleaseSourceVariant(releasePayload: ReleasePayload): ReleaseVariant {
  const variants = extractReleaseSourceVariants(releasePayload);
  if (variants.length === 0) throw new Error("no_supported_torrent_variant");

  for (const preferredCodec of ["avc", "hevc"] as const) {
    const preferred = variants.filter(v => v.codec === preferredCodec && v.magnet && v.available_episodes.length > 0);
    if (preferred.length > 0) {
      preferred.sort((a, b) => {
        const maxA = Math.max(...a.available_episodes);
        const maxB = Math.max(...b.available_episodes);
        if (maxB !== maxA) return maxB - maxA;
        if (b.available_episodes.length !== a.available_episodes.length) return b.available_episodes.length - a.available_episodes.length;
        return (a.resolution ?? "").localeCompare(b.resolution ?? "") || (a.label ?? "").localeCompare(b.label ?? "");
      });
      return preferred[0]!;
    }
  }

  throw new Error("no_supported_torrent_variant");
}

export function extractReleaseTitle(releasePayload: ReleasePayload): string {
  const names = (releasePayload.name ?? releasePayload.names ?? {}) as Record<string, string>;
  if (typeof names === "object") {
    for (const key of ["english", "en", "main", "ru", "alternative"]) {
      const v = names[key];
      if (typeof v === "string" && v.trim()) return v.trim();
    }
  }
  if (typeof releasePayload.title === "string" && releasePayload.title.trim()) return releasePayload.title.trim();
  if (typeof releasePayload.alias === "string" && releasePayload.alias.trim()) return releasePayload.alias.trim();
  throw new Error("missing_title");
}

export function extractReleaseTitleRu(releasePayload: ReleasePayload): string | null {
  const names = (releasePayload.name ?? releasePayload.names ?? {}) as Record<string, string>;
  if (typeof names === "object") {
    for (const key of ["main", "ru"]) {
      const v = names[key];
      if (typeof v === "string" && v.trim()) return v.trim();
    }
  }
  return null;
}

export function extractReleaseSeason(releasePayload: ReleasePayload): number {
  const season = releasePayload.season_number ?? releasePayload.seasonNumber;
  const parsed = Number(season);
  return Number.isNaN(parsed) || parsed < 1 ? 1 : parsed;
}

export function extractReleaseMalId(releasePayload: ReleasePayload): number | null {
  const directKeys = ["mal_id", "malId", "myanimelist_id", "myanimelistId"];
  for (const key of directKeys) {
    const v = (releasePayload as Record<string, unknown>)[key];
    const parsed = parsePositiveInt(v);
    if (parsed !== null) return parsed;
  }

  const nestedCandidates: [string, string][] = [
    ["external_ids", "mal_id"], ["external_ids", "malId"], ["external_ids", "myanimelist"],
    ["external_ids", "myanimelist_id"], ["externalIds", "mal_id"], ["externalIds", "myanimelist"],
    ["codes", "mal"], ["codes", "mal_id"], ["player", "mal_id"], ["player", "myanimelist"],
    ["metadata", "mal_id"],
  ];
  for (const [parentKey, childKey] of nestedCandidates) {
    const parent = (releasePayload as Record<string, unknown>)[parentKey];
    if (typeof parent !== "object" || parent === null) continue;
    const parsed = parsePositiveInt((parent as Record<string, unknown>)[childKey]);
    if (parsed !== null) return parsed;
  }
  return null;
}

function parsePositiveInt(value: unknown): number | null {
  const parsed = Number(value);
  return Number.isNaN(parsed) || parsed <= 0 ? null : parsed;
}
